### ISSUE TITLE
Avoid infinite recursion during object building from the parameter file

### ISSUE THREAD

**Issue by abensonca** [galacticusorg/galacticus#397]
Objects which composite a member of their own class can get into an infinite recursion if no such object is explicitly provided to them (they instead try to composite themselves). Add an option to the `objectBuilder` directive (or maybe even automate this) that indicates that an object of the class must be explicitly provided to avoid this problem.

**Comment by abensonca** [comment]
A good example of this would be in `source/galactic.filters.not.F90`. This file implements a meber of the `galacticFilter` function class:
```
  <galacticFilter name="galacticFilterNot">
   <description>Implements logical negation of a wrapped \refClass{galacticFilterClass}, passing nodes that the wrapped filter would reject and rejecting those it would pass, enabling complement-based selection criteria without defining explicit inverse filters.</description>
  </galacticFilter>
```
In its constructor `notConstructorParameters` it then has:
```
    <objectBuilder class="galacticFilter" name="galacticFilter_" source="parameters"/>
```
which builds a member of this same function class.

Given a parameter file containing:
```
<galacticFilter value="not">
  <galacticFilter value="haloIsolated"/>
</galacticFilter>
```
this succeeds, as the `haloIsolated` case is picked up from inside the `<galacticFilter value="not">` block. But, given:
```
<galacticFilter value="not"/>
```
this will fail - there is no nested `<galacticFilter>` element, so the search works up the tree, finding the original `<galacticFilter value="not"/>` which is then attempts to build, leading to an infinite recursion.

**Comment by abensonca** [comment]
# Investigation: #397 — Avoid infinite recursion during object building from the parameter file

**Status: investigation only — no source changes.** All findings below were verified against current `master` (build system: `python/Galacticus/Build/SourceTree/Process/ObjectBuilder.py`; `Galacticus.exe` built 13-June-2026).

## TL;DR

The headline claim — that `<galacticFilter value="not"/>` triggers *infinite* recursion — **is no longer accurate on current `master`.** An existing recursion guard (emitted by `ObjectBuilder.py`, present since the Python port on 2026-04-24, i.e. *before* this issue was filed on 2026-06-12) catches the simple self-composition case and aborts with a fatal error rather than recursing. I reproduced this empirically (details below).

What remains valid in the issue is the underlying request: the current behaviour is a **reactive runtime safety net with a cryptic message and real coverage gaps**, not a clean, declarative "this object must be provided explicitly" mechanism. That feature is still worth adding.

---

## 1. Reproduction

### Minimal failing case (the issue's example)

Parameter fragment placed where a `galacticFilter` is consumed (here, the standard merger-tree outputter):

```xml
<mergerTreeOutputter value="standard">
  <galacticFilter value="not"/>   <!-- empty: no wrapped filter provided -->
</mergerTreeOutputter>
```

Exact recipe used (so it can be re-run):
- Base file: `testSuite/parameters/benchmark_milkyWay.xml` (small, 2 trees, fixed mass), with the outputter block above injected and `outputFileName` redirected to scratch.
- Run: `OMP_NUM_THREADS=1 ./Galacticus.exe <paramfile>` (the `xi:include`s are path-relative, so the file must live under `testSuite/parameters/`).

**Observed result (current `master`): a fatal error during initialisation — NOT a hang or stack overflow.** Exit code 1, with backtrace:

```
#1 __error_MOD_error_report_char            error.F90:185
#2 __galactic_filters_MOD_notconstructorparameters   galactic.filters.not.F90:61
#3 galacticfiltercnstrctrprmtrs             galactic.filters.F90:2696
#4 standardconstructorparameters            merger_trees.outputter.standard.F90:121
#5 mergertreeoutputtercnstrctrprmtrs        merger_trees.outputter.F90:897
#6 evolveforestsconstructorparameters       tasks.evolve_forests.F90:579
...
Fatal error:
recursive build of [galacticFilter] class detected
 Occurred at: directive:objectBuilder  function:notConstructorParameters  file:galactic.filters.not.F90 [line 49]
```

### Control: valid nested case still works

```xml
<mergerTreeOutputter value="standard">
  <galacticFilter value="not">
    <galacticFilter value="mainBranch"/>
  </galacticFilter>
</mergerTreeOutputter>
```

Runs to completion (exit 0, HDF5 written). So the guard does not break legitimate usage.

---

## 2. Root cause

### The object-build search

For an `objectBuilder` directive with no `parameterName` (the default-named case, where the parameter key equals the class name — e.g. `galacticFilter`), `ObjectBuilder.py` emits a search that walks **up** the parameter tree until it finds the named parameter (`ObjectBuilder.py:164-170`; generated in `galactic.filters.not.p.F90:52-55`):

```fortran
parametersCurrent => parameters
do while (.not.parametersCurrent%isPresent('galacticFilter').and.associated(parametersCurrent%parent))
   parametersCurrent => parametersCurrent%parent
end do
```

When `galacticFilterNot`'s constructor runs, `parameters` is the sub-parameter set scoped to the `<galacticFilter value="not">` element (`inputParametersSubParameters` sets `subParameters%parent => self`, `utility.input_parameters.F90:2092`). For an *empty* `not` block there is no nested `galacticFilter`, so the loop steps up exactly one level to `parameters%parent` — the node that *contained* the `<galacticFilter value="not"/>` element. That ancestor, by construction, **does** contain a `galacticFilter` element (the very `not` we are inside). Without further protection, the constructor would then rebuild that same `not` element → unbounded recursion. This is exactly the mechanism the issue describes.

### Why it does not actually recurse today: the guard

Because `self` is in scope and is of the target class, `ObjectBuilder.py:189-199` additionally emits a recursion guard (`galactic.filters.not.p.F90:56-62`):

```fortran
if (parametersCurrent%isPresent('galacticFilter')) then
   genericObject => self
   select type (genericObject)
   class is (galacticFilterClass)
      if (associated(parametersCurrent,parameters%parent)) &
         call Error_Report('recursive build of [galacticFilter] class detected'//...)
   end select
   ...
```

In the failing case the loop lands precisely on `parameters%parent`, so `associated(parametersCurrent, parameters%parent)` is `.true.` and the build aborts cleanly. For same-class self-composition this is robust at any nesting depth, because the walk-up can only ever stop at the immediate containing element (which is always the recursion source).

**So the precise failure on `master` is: a fatal `Error_Report`, not infinite recursion.** The defect that remains is one of *ergonomics and coverage*, not a live hang.

---

## 3. Scope

### How often this bites
- Triggered whenever a class that composites a member of **its own functionClass** via a *default-named* `objectBuilder` is instantiated **without** the inner object being provided.
- Self-compositing `galacticFilter` subclasses are common: `not`, `parentNode`, `childNode`, `descendantNode`, `primaryDescendantNode`, `anyDescendantNode`, `anySatelliteNode`, `isolatedHostNode`, `null`, plus the list-style `all`/`any`. The same shape exists in other hierarchies (e.g. `darkMatterProfileScaleRadius` wrappers).
- It is a user-input error (omitting a required nested object), so it surfaces at model setup time, before evolution.

### What the guard covers — and three gaps
1. **Covered:** default-named, same-class composition where `self` is declared and is of the class. (Verified.)
2. **Gap — cryptic message.** "recursive build of [galacticFilter] class detected" tells the user *what the engine detected*, not *what they did wrong*. Compare the hand-written check in `nodes.operators.filtered.F90:63`: `"A [galacticFilter] must be given explicitly for this class"` — clearer, and exactly the behaviour the issue asks to generalise.
3. **Gap — `self` must be in scope and of the class.** The guard is gated on `declaration_exists(parent,'self')` and a `select type ... class is (<class>Class)`. A constructor whose result variable is not named `self` would get no guard. *(Reasoned from the code; not separately reproduced.)*
4. **Gap — mutual recursion is NOT caught.** If class A composites class B (default-named) and B composites A, then inside A's constructor `self` is an A (not a B), so the `class is (BClass)` arm never matches and no guard is emitted for that build; likewise in B. Such an A↔B pair could still recurse without bound. *(This is a reasoned hypothesis from reading the guard logic; I did **not** find or construct a concrete A↔B pair in the tree to confirm one exists in practice — evidence here is incomplete.)*

### Related, intentional mechanism
Classes that *legitimately* composite their own class already opt in via `recursive="yes"` on the class directive (e.g. `darkMatterHaloScaleVirialDensityContrastDefinition`, `virialDensityContrastPercolation`, `starFormationHistoryFixedAges`). The factory then breaks the recursion deliberately using `isRecursive`/`recursiveSelf` pointers (`FunctionClass/__init__.py:3001-3159`) instead of re-entering the constructor. This is the existing precedent for a *declarative* recursion attribute and is the natural place to hang a complementary "must be explicit" flag.

---

## 4. Options

> Note on numerical / build behaviour: none of these change results for *valid* parameter files — they only change the failure mode for invalid (omitted-object) input. Options that touch generated code will force recompilation of affected source (build-time cost only). No change to physics or numerics is expected in any option.

### Option A — Add an opt-in `objectBuilder` attribute (matches the issue's primary suggestion)
Add e.g. `requireExplicit="yes"` to the `objectBuilder` directive (schema `schema/objectBuilder.xsd`, handler `ObjectBuilder.py`). When set, emit a no-walk-up presence check with a clear message instead of (or before) the walk-up search:

```fortran
if (.not.parameters%isPresent('galacticFilter')) &
   call Error_Report('a [galacticFilter] must be provided explicitly'//<loc>)
```

- **Pros:** Minimal blast radius (only files that add the attribute change). Clear message. Generalises the existing hand-rolled `filtered` idiom. Directly fulfils the request.
- **Cons:** Opt-in — every self-compositing class must be annotated, so coverage depends on remembering to set it. Does not, by itself, fix mutual recursion unless applied on both sides.
- **Risk:** **Low.**

### Option B — Automate detection at build time
In `ObjectBuilder.py`, when a default-named build's `class` equals the enclosing functionClass and that class is **not** `recursive="yes"`, automatically emit the require-explicit check (Option A's code) instead of relying solely on the runtime guard.

- **Pros:** No per-class annotation; closes the whole same-class category at once with a good message. The `recursive="yes"` set provides a ready-made allowlist of legitimate exceptions. Safe for valid parameter files (a same-class walk-up can only ever resolve to the recursion source, so nothing legitimate is lost — this is why the existing guard is universally correct).
- **Cons:** Changes generated code for many constructors → broad recompilation. Requires the directive processor to reliably know "the enclosing functionClass" for each `objectBuilder` (the `self`/class-name introspection already used by the guard is the obvious basis). Still does not address cross-class mutual recursion.
- **Risk:** **Low–Medium** (mostly build-time scope and getting the "enclosing class" detection right).

### Option C — Keep the runtime guard, just improve it
Leave the architecture as-is but (i) reword the message to the "must be provided explicitly" form, and (ii) optionally broaden the guard to a generic build-stack check that also catches mutual recursion (e.g. a small thread-private stack of `(class, parameterNode)` currently being built, error on repeat).

- **Pros:** Smallest change for (i); (ii) is the only option here that addresses mutual recursion. Stays reactive/runtime so it needs no per-class knowledge.
- **Cons:** (i) alone is cosmetic and stays reactive (fails at the recursion point, not at the point of omission). (ii) adds per-build bookkeeping in a hot setup path and is more invasive.
- **Risk:** **Low** for (i); **Medium** for (ii).

### Recommendation
**Option B as the primary fix, with Option C(i)'s message wording folded in.** Automating the same-class case removes the foot-gun for the entire (large) family of self-compositing classes without per-class annotation, keeps `recursive="yes"` as the explicit "this one is intentional" escape hatch, and is safe for all valid input. Keep **Option A** available as the explicit attribute for the rarer cases the automatic rule can't infer (e.g. `parameterName`-distinct builds that are nonetheless same-hierarchy). Treat **mutual recursion (Option C-ii)** as a separate, lower-priority follow-up — first confirm whether any real A↔B pair exists before investing in a build-stack guard.

---

## Evidence ledger
- **Verified empirically:** simple `<galacticFilter value="not"/>` → fatal "recursive build detected" error, no infinite recursion (exit 1, backtrace above); valid nested `not`/`mainBranch` → completes (exit 0).
- **Verified by source reading:** walk-up search and guard generation (`ObjectBuilder.py:119-199`); generated output (`work/build/galactic.filters.not.p.F90:52-95`); `subParameters%parent` linkage (`utility.input_parameters.F90:2092`); hand-written precedent (`nodes.operators.filtered.F90:63`); `recursive="yes"` mechanism (`FunctionClass/__init__.py:3001-3159`); directive schema (`schema/objectBuilder.xsd`).
- **Not verified / incomplete evidence:** existence of a concrete cross-class mutual-recursion pair; behaviour for constructors lacking a `self` variable. Both are reasoned from the guard's gating conditions, not reproduced.

**Comment by abensonca** [comment]
## Follow-up: confirmed mutual recursion is reachable, and the `functionGlobal` loophole

Two further findings while scoping whether a mutual-recursion guard is actually needed (vs. just an improved message for the simple self-composition case).

### 1. Cross-class mutual recursion is real and reachable (reproduced)

Fortran's module system mostly protects us: if class A (module `M_A`) composites class B, `objectBuilder` emits `use M_B`; the reverse needs `use M_A` → circular module `use` → won't compile. So cross-**module** mutual recursion can't be built.

The exception is **two functionClasses sharing one module**. There are 5 such modules; only `structure_formation.cosmological_density_field.F90` actually contains a cycle. It hosts `criticalOverdensity`, `haloEnvironment`, and `cosmologicalMassVariance`, which cross-composite:
- `criticalOverdensityEnvironmental` builds a `haloEnvironment` (`...critical_overdensity.environmental.F90:81`)
- `haloEnvironmentNormal`/`LogNormal` build a `criticalOverdensity` (`...halo_environment.normal.F90:118`, `lognormal.F90:91`)

All default-named `objectBuilder`s, no explicit guards. The existing same-class recursion guard **cannot** fire because in each constructor `self` is a *different* class than the one being built.

A parameter file with two mutually-resolving siblings triggers it (neither's need satisfied by an inner child — the environmental crit's inner terminal `criticalOverdensity` is a *grandchild* of the level `haloEnvironment` walks up to, so it's missed):

```xml
<criticalOverdensity value="environmental">
  <criticalOverdensity value="sphericalCollapse..."/>  <!-- satisfies its OWN same-class build -->
</criticalOverdensity>
<haloEnvironment value="normal"> ... </haloEnvironment>  <!-- sibling -->
```

This **segfaults** (SIGSEGV, exit 139) at initialization with no error message. gdb shows a clean, infinitely-repeating 4-frame cycle:

```
environmentalConstructorParameters   (criticalOverdensity = environmental)
  → haloEnvironment factory
  → normalConstructorParameters       (haloEnvironment = normal)
  → criticalOverdensity factory
  → environmentalConstructorParameters … repeats until stack overflow
```

(Normal/working configs avoid this by giving the top-level `haloEnvironment` a *terminal* `criticalOverdensity` sibling, e.g. `concentrationDistributionLudlow2016Environmental.xml:54`. Nothing stops a user writing the recursive variant.)

So a guard scoped only to *same-class* composition is insufficient — a general build-stack guard is justified.

### 2. The `functionGlobal` loophole

`functionGlobal` can subvert the module-hierarchy protection above. It declares a procedure **pointer** `<name>_` in a low-level module (defaulting to a `_Null` stub), wired at runtime to the real function by an `establish` pass. Any module can then call through the pointer **without** a `use` of the target's module. The intent is stated verbatim in `nodes.operators.utilities.F90:51`: *"a globally-callable function to allow us to subvert the class/module hierarchy."*

Two **constructors** are exposed this way, each with one current caller:

| Global constructor | Called by |
|---|---|
| `nodeOperatorConstruct_` | `geometry.lightcones.square.F90:286` (`geometryLightconeSquare`) |
| `accretionDiskSpectraConstruct_` | `radiation.intergalactic_background.internal.F90` |

So e.g. `geometryLightconeSquare` builds a `nodeOperator` despite sitting below `Nodes_Operators` in the module order — exactly the kind of "lower module constructs a higher class" path that could, in principle, close a recursion the compile-time barrier would otherwise block.

**Would a build-stack guard catch a recursion through this path? Yes.** The global wrapper is a thin shim that delegates to the *normal generated factory*: `nodeOperatorConstruct` ends in `nodeOperator__ => nodeOperator(parametersCurrent)` (`nodes.operators.utilities.F90:66`) — the same `*cnstrctrprmtrs` factory that appeared in the mutual-recursion backtrace above. Both the ordinary `objectBuilder` path and the `functionGlobal` path re-enter that factory, so a guard instrumented there fires regardless of how the constructor was reached.

**Caveat — guard placement matters.** The wrapper does its own hand-written walk-up (`nodes.operators.utilities.F90:62-65`) and is not generated by `objectBuilder`. So a "must be explicit" check emitted only at `objectBuilder` expansion sites would **not** cover the wrapper; a **factory-level** build-stack guard would. This argues for placing the guard in the generated factory dispatch rather than (only) at `objectBuilder` sites.

**Is a concrete cycle reachable through `functionGlobal` today?** Not that I found. A `nodeOperator`↔`geometryLightcone` cycle would need some `nodeOperator` to composite a `geometryLightcone`; none does (the 7 `geometryLightcone` composers are timestep/operator/extractor/filter/SFH classes, no `nodeOperator`). I did not trace transitive/user-config-induced cycles, so I can't fully rule one out, but there's no evidence of a currently-reachable crash via this path — unlike the `criticalOverdensity`↔`haloEnvironment` cycle, which reproduces.

### Net
- A **factory-level build-stack guard** is the robust fix: it catches both the module-internal cross-cycle (reproduced) and any `functionGlobal`-mediated path, and is independent of whether the constructor was reached via `use` or a global pointer.
- An improved error message for the common *self*-composition case (`galacticFilter value="not"`, etc.) remains a cheap, independent win on top of that.
